File: scala/bob/src/main/scala/Bob.scala
/**
  * Created by Angelo on 9/6/2016.
  */

class Bob {
  def hey(s: String) : String = {
    if (s.trim().length() == 0) return  "Fine. Be that way!"
    if (s.forall(x => (x.isUpper || !x.isLetter) && s.count(x => x.isLetter) > 0)) return "Whoa, chill out!"
    if (s.endsWith("?")) return "Sure."
    "Whatever."
  }
}

File: fsharp/bob/Bob.fs
module Bob

type Address = Statement | Question | Exclaimation | Silence

let determineTypeOfAddress (address:string) = 
    match address with
    | s when System.String.IsNullOrWhiteSpace(address) -> Silence
    | s when 
        let allLetters = address.ToCharArray() |> Array.toList |> List.filter (fun x -> System.Char.IsLetter(x)) 
        allLetters 
        |> List.length > 0 
        && allLetters 
        |> List.forall (fun x -> System.Char.IsUpper(x)) -> Exclaimation
    | s when address.EndsWith("!") -> Exclaimation
    | s when address.EndsWith("?") -> Question
    | _ -> Statement

type Bob(address) = 
    let incoming = determineTypeOfAddress address
    let respond address =
        match address with
        | Statement -> "Whatever."
        | Question -> "Sure."
        | Exclaimation -> "Woah, chill out!"
        | Silence -> "Fine, be that way."
    member this.hey() = 
        respond incoming